### ISSUE TITLE
Can we get a middle 'operation' tensor value like it in TF1?

### ISSUE THREAD

**Issue by JWenBin** [tensorflow/java#199]
We trainned our model in TensorFlow, after that, use it in spark. And we can get middle value like:
**TensorFlow1.x (python):**
...
mid_out = tf.nn.relu(y_deep,name="mid_out")  
...
final_out = tf.nn.softmax(tf.add(tf.matmul(concat_input, weights['concat_projection']), weights['concat_bias']),name="final_out")  
...
**spark:**
val z = sess.runner()
      .feed("input_layer", input_layer)
      .fetch("mid_out")
      .run()
      .get(0)

But now, I find it different in **TensorFlow2**+: I must add 'serving_default_' in feed(),  and I could only use ‘StatefulPartitionedCall’ instead of ‘name final_out’ to get the predict result:
  val z = sess.runner()
      .feed("**serving_default_**input_layer", input_layer)
      .fetch("**StatefulPartitionedCall**")
      .run()
      .get(0)
**In fact, I need the middle value 'mid_out' too, but I don't know how to get it like it in TF1.X.**

Any tips will be great,  thanks!

**Comment by rnett** [comment]
The Java APIs work the same, I'd guess the graph generated by your python code has changed.  I'd start by inspecting it, or sharing it here if you can.  It probably has to do with the SavedModel format changes, see https://www.tensorflow.org/guide/saved_model and https://www.tensorflow.org/tfx/serving/signature_defs#signaturedef_structure

What are you using to export and load your model?

**Comment by JWenBin** [comment]
> 
> 
> The Java APIs work the same, I'd guess the graph generated by your python code has changed. I'd start by inspecting it, or sharing it here if you can. It probably has to do with the SavedModel format changes, see https://www.tensorflow.org/guide/saved_model and https://www.tensorflow.org/tfx/serving/signature_defs#signaturedef_structure
> 
> What are you using to export and load your model?

Thank you for your help! Yes, I changed my python code. In TF1.X, I used:
    frozen_graph_def = tf.graph_util.convert_variables_to_constants(
        sess,
        sess.graph_def,
        ["input_layer","mid_out","final_out"])
    with open(model_path+'YouTubeNet_model.**pb**', 'wb') as f:
        f.write(frozen_graph_def.SerializeToString())

And in TF2.X, I used:
**tf.saved_model.save**(model, outputDir+'YouTubeNet_model')
Should I frozen the graph first with TF2.X instead of using tf.saved_model.save directly?

Thanks!

**Comment by JWenBin** [comment]
**And I use the way to load my model like this:**
import org.{tensorflow=>tf}
val bundle = tf.SavedModelBundle.load("d:/data\\output\\YouTubeNet\\YouTubeNet_model","serve")
val sess = bundle.session()

**Comment by rnett** [comment]
I'm assuming you're using `tf.Module` style model declarations?  As best I can tell, tensorflow exports the `@tf.function` functions as functions in the graph, and then calls them.  So there's no global `mid_out` op, it's only in the scope of the function, like a local variable in Java.  I don't think there's anything we can do about that on the Java side (cc @karllessard).

If you can freeze and save the model in the same way you did in tf1, it's worth a shot.  [This SO question](https://stackoverflow.com/questions/62984496/fetching-an-op-from-a-savedmodel-in-tf2) seems to be for the same issue, but has no answers.

**Comment by JWenBin** [comment]
Thanks! May be I will use TF1.X instead of TF2.x for quite a long time.
By the way, I use tf.keras.models.Model which I think is inherited from tf.Module in TF2.X.

**Comment by karllessard** [comment]
Hi @JWenBin , can you try calling `tf.saved_model.save` but passing the signature explicitly, like described in [this post](https://github.com/tensorflow/tensorflow/issues/32674#issuecomment-559386388)?

**Comment by JWenBin** [comment]
> 
> 
> Hi @JWenBin , can you try calling `tf.saved_model.save` but passing the signature explicitly, like described in [this post](https://github.com/tensorflow/tensorflow/issues/32674#issuecomment-559386388)?

**Good idea! I tried as you say and success! Thank you very much!**
By the way, my Python code is (TF2):
......
    full_model = tf.function(lambda Input: model(Input))
    full_model = full_model.get_concrete_function([tf.TensorSpec(model.inputs[0].shape, model.inputs[0].dtype),tf.TensorSpec(model.inputs[1].shape, model.inputs[1].dtype),tf.TensorSpec(model.inputs[2].shape, model.inputs[2].dtype)])
    **tf.saved_model.save(model, outputDir+'YouTubeNet_model',signatures=full_model)**
......
And then, I can get the 'mid_out' by using ‘ .fetch("StatefulPartitionedCall:Index")’ successfully!

**Comment by JWenBin** [comment]
I also tried to freeze and save the model in the same way I did in TF1, and reload it on my **spark driver** successfully.
    val graph = new Graph()
    val graphBytes = IOUtils.toByteArray(new FileInputStream(pbtxtModelPath ))
    graph.importGraphDef(graphBytes);
    val session: Session = new Session(graph)

In order to reload the model on spark executors, I do not need to freeze it anymore, and we can reload the model as before.
    val bundle = tf.SavedModelBundle.load(pbtxtModelPath,"serve")
    val sess = bundle.session()
